presto-raptor/src/main/java/com/facebook/presto/raptor/storage/OrcStorageManager.java

```diff
@@ -457,6 +516,79 @@ private ShardInfo createShardInfo(UUID shardUuid, OptionalInt bucketNumber, Path
         return shardDelta(shardUuid, Optional.of(shard));
     }
 
+    @VisibleForTesting
+    Collection<Slice> writeDeltaDeleteFile(long transactionId, OptionalInt bucketNumber, UUID oldShardUuid, Optional<UUID> oldDeltaDeleteShardUuid, BitSet rowsToDelete)
+    {
+        if (rowsToDelete.isEmpty()) {
+            return ImmutableList.of();
+        }
+        // blockToDelete is LongArrayBlock
+        List<Long> columnIds = new ArrayList<>();
+        List<Type> columnTypes = new ArrayList<>();
+        columnIds.add(0L);
+        columnTypes.add(BIGINT);
+
+        // todo one block can't hold one orcfile
+        // At this point, rowsToDelete couldn't be empty
+        boolean oldShardContainToDelete = false;
+        if (oldDeltaDeleteShardUuid.isPresent()) {
+            oldShardContainToDelete = mergeRowsToDelete(rowsToDelete, oldDeltaDeleteShardUuid.get());
+        }
+
+        if (rowsToDelete.cardinality() == getRowCount(oldShardUuid)) {
+            // Delete original file
+            return shardDeleteDelta(oldShardUuid, oldDeltaDeleteShardUuid, Optional.empty());
+        }
+        if (oldShardContainToDelete) {
+            return ImmutableList.of();
+        }
+
+        BlockBuilder blockBuilder = new LongArrayBlockBuilder(null, rowsToDelete.size());
+        for (int i = rowsToDelete.nextSetBit(0); i >= 0; i = rowsToDelete.nextSetBit(i + 1)) {
+            blockBuilder.writeLong(i);
+        }
+        Block blockToDelete = blockBuilder.build();
```

Comment:
In the read path of delta you allow a delta to have multiple blocks. But here in the write path we always try to use one block to hold all row ids and flush it at one time. Maybe add some comments around this behavior.